### PR TITLE
Add an endpoint for get pickersets

### PR DESCRIPTION
**Pull request by 70ray**
We could avoid the expand_pickersets() function by creating the pickersets with character names to start with but then we would have to change some other stuff to use the data in this form.

Sandbox at: https://www.pgdp.org/~rp31/c.branch/api_pickers

### CODE REVIEW

File: api/v1_projects.inc [hunk 1]
```diff
@@ -799,6 +799,40 @@ function api_v1_project_validatetext($method, $data, array $query_params)
...
 }
 
+function api_v1_project_pickersets($method, $data, $query_params)
+{
+    // expand the character ranges so that each character can have its name
+    $project = $data[":projectid"];
+    return expand_pickersets($project->get_pickersets());
+}
+
+function expand_pickersets($pickersets)
```
Comment (cpeel) [inline api/v1_projects.inc:809]: Lets put most of this functionality in the PickerSet class, maybe a `get_verbose_subsets()` (or a better-named) function, and then move the API-specific bits like mapping to an array into `api_v1_project_pickersets()` directly.

Comment (70ray) [reply api/v1_projects.inc]: Done something like this.

File: pinc/CharSuites.inc [hunk 2]
```diff
@@ -31,6 +31,33 @@ class PickerSet
...
+                    }
+                }
+                $verbose_rows[] = $char_names;
+            }
+            $verbose_subsets[$subset_name] = $verbose_rows;
+        }
+        return $verbose_subsets;
+    }
+
+    public function get_titles(): array
```
Comment (jchaffraix) [inline pinc/CharSuites.inc:56]: Thanks for adding a return type to this function!

To help PHPStan, we should add the precise type when it is easy to infer/simple:

```suggestion
    /** @return string[] */
    public function get_titles(): array
```

**Review by jchaffraix — APPROVED** [review]

File: api/v1_projects.inc [hunk 3]
```diff
@@ -799,6 +799,21 @@ function api_v1_project_validatetext($method, $data, array $query_params)
     return $text_checker->analyse(receive_project_text_from_request_body('text'), $data[":projectid"]);
 }
 
+function api_v1_project_pickersets($method, $data, $query_params)
+{
+    $project = $data[":projectid"];
+    $verbose_pickersets = [];
+    $pickersets = $project->get_pickersets();
+    foreach ($pickersets as $pickerset) {
```
Comment (cpeel) [inline api/v1_projects.inc:807]:
```suggestion
    foreach ($project->get_pickersets() as $pickerset) {
```

File: api/v1_projects.inc [hunk 4]
```diff
@@ -799,6 +799,21 @@ function api_v1_project_validatetext($method, $data, array $query_params)
...
+        $verbose_pickersets[] = [
+            "name" => $pickerset->name,
+            "subsets" => $pickerset->get_verbose_subsets(),
+            "titles" => $pickerset->get_titles(),
+        ];
```
Comment (cpeel) [inline api/v1_projects.inc:812]: Would the return value for this be more useful to the client if it looked like:
```
$verbose_pickersets[] = [
    "name" => $pickerset->name,
    "subsets" =
        [
            "title" => "string",
            "characters" => [...],
        ],
        [
            "title" => "string2",
            "characters" => [...],
        ],
];
```
Or something similar? That way the client doesn't have to map the titles back to the subsets?

**Review by cpeel — COMMENTED** [review]
Having this in the PickerSet class makes a lot more sense, thanks for moving it in there.

File: api/v1_projects.inc [hunk 5]
```diff
@@ -799,6 +799,21 @@ function api_v1_project_validatetext($method, $data, array $query_params)
     return $text_checker->analyse(receive_project_text_from_request_body('text'), $data[":projectid"]);
 }
 
+function api_v1_project_pickersets($method, $data, $query_params)
```
[the remark this answers is not in the document]
Comment (70ray) [reply api/v1_projects.inc:802]: Yes indeed.

File: api/v1_projects.inc (hunk 4 above)
Comment (70ray) [reply api/v1_projects.inc:812]: Yes, thta would be better

File: pinc/CharSuites.inc (hunk 2 above)
Comment (70ray) [reply pinc/CharSuites.inc:56]: removed this function now.

**Review by jchaffraix — APPROVED** [review]

**Comment by 70ray** [comment]
Sandbox updated.

**Review by cpeel — APPROVED** [review]
Brilliant.

For other reviewers, this can be tested with:
```
export API_KEY=review
curl -i -H "Accept: application/json" -H "X-API-KEY: $API_KEY" \
     -X GET "https://www.pgdp.org/~rp31/c.branch/api_pickers/api/v1/projects/projectID45b5cfaa0ff81/pickersets"
```

**Comment by srjfoo** [comment]
> Brilliant.
> 
> For other reviewers, this can be tested with:
> 
> ```
> export API_KEY=review
> curl -i -H "Accept: application/json" -H "X-API-KEY: $API_KEY" \
>      -X GET "https://www.pgdp.org/~rp31/c.branch/api_pickers/api/v1/projects/projectID45b5cfaa0ff81/pickersets"
> ```

Thanks, Casey, that helps immensely.

@70ray, in looking at the output, the thing that stood out to me is that the numbers don't show the actual character. They're also bracketed by square brackets instead of curly, as with the others. I have a memory of you fixing something similar to that in the code -- is that what's going on here? (see screenshot below of picker set labeled `1`.)

image

**Comment by cpeel** [comment]
> @70ray, in looking at the output, the thing that stood out to me is that the numbers don't show the actual character. They're also bracketed by square brackets instead of curly, as with the others. I have a memory of you fixing something similar to that in the code -- is that what's going on here? (see screenshot below of picker set labeled `1`.)

Oh, very good catch Sharon! We need to massage those into a string type instead of an int type. I forget exactly how we do that but you are correct we do that elsewhere.

**Comment by 70ray** [comment]
> > @70ray, in looking at the output, the thing that stood out to me is that the numbers don't show the actual character. They're also bracketed by square brackets instead of curly, as with the others. I have a memory of you fixing something similar to that in the code -- is that what's going on here? (see screenshot below of picker set labeled `1`.)
> 
> Oh, very good catch Sharon! We need to massage those into a string type instead of an int type. I forget exactly how we do that but you are correct we do that elsewhere.

An easy work-around would be to return pairs [character, name] rather than an assciative array. This would be just as easy for the client to use. So the result would be:
image

**Comment by cpeel** [comment]
> An easy work-around would be to return pairs [character, name] rather than an assciative array. This would be just as easy for the client to use.

We could, but we're doing something to coerce some array key numbers into strings already if you look at the very top of Sharon's screenshot: 
image

What are we doing that coerces the `1`?

**Comment by 70ray** [comment]
> > An easy work-around would be to return pairs [character, name] rather than an assciative array. This would be just as easy for the client to use.
> 
> We could, but we're doing something to coerce some array key numbers into strings already if you look at the very top of Sharon's screenshot: image
> 
> What are we doing that coerces the `1`?

It seems that "1" is because the keys are a mixture or numbers and other characters, but if you use consecutive numbers starting from 0 it gets turned into an ordinary array. But we can coerce it into an object with (object).

**Comment by 70ray** [comment]
But it seems an object is not the same sort of thing as an associative array.

**Comment by 70ray** [comment]
Test has been adjusted so it works with an object. Is this a good solution or would returning the [char, name] pairs be better? Or if we went back to the state of [93e73a3] I think it would work ok since the keys of the simple array would be the correct numbers and javascript would tun them back into strings.

**Comment by 70ray** [comment]
Sandbox updated.

**Review by srjfoo — APPROVED** [review]

**Comment by cpeel** [comment]
> Is this a good solution or would returning the [char, name] pairs be better? I like the pairs better.

I think this solution works well. I'm partial to the hash as I think it lets clients easily get the keys via built-in functions. I'm not against the pairs though if you'd rather do that -- both seem acceptable to me. Whichever you prefer is fine.

**Comment by 70ray** [comment]
I looked at how this appears when it gets to the javascript. Unfortunately the "1: numbers and math" key appears as an integer so the subsets get in the wrong order. [mdn](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...in) says:
"Within each component of the prototype chain, all non-negative integer keys (those that can be array indices) will be traversed first in ascending order by value, then other string keys in ascending chronological order of property creation."
I think the best solution would be to treat the components that we want to keep in the right order as simple arrays. I'll made it like this.

**Comment by 70ray** [comment]
Github doesn't seem to be running the checks

**Comment by 70ray** [comment]
Conflicts resolved, snadbox updated.

**Review by cpeel — APPROVED** [review]
This LGTM!

**Review by srjfoo — APPROVED** [review]

**Pull request merged**